### DISCUSSION TITLE
MicroMod to Development branch

### DISCUSSION THREAD

**Discussion by Satnet** [luni64/TeensySharp#8] [Ideas]
Request to add Micromod Teensy to the Development branch and then push the branch to master.

**Comment by luni64** [comment]
Currently working on it

**Reply by Satnet** [reply]
Wonderful, will it also be pushed to master with examples as the code seems to be quite different from the previous version?

**Reply by luni64** [reply]
Sure, but I would need someone to test before pushing to master. Maybe you find some time? Hope to finish it on the weekend.

**Reply by Satnet** [reply]
I'd love to test it and help out as much as I can, but this is kind of new teritory for me with USB and stuff. Let me know if I can do anything. I can test it for you if you can tell me what files to include, there are allot of files in different folders such as Native and stuff. :)

**Reply by luni64** [reply]
>I'd love to test it and help out as much as I can

Ok, I pushed the underlying libUsbTree  library to nuget and changed the TeensySharp project to download it from nuget.  The detection part should work quite well, the downloading is not very stable at the moment. 

Anyway, can you clone the repository, check out the development branch and see if the print&watch example compiles and runs? 
It should display a list of all connected USB devices. The Teensies should be printed in red. Connecting / disconnecting should generate corresponding messages. 

![image](https://user-images.githubusercontent.com/12611497/125206888-b7624600-e289-11eb-9d03-e4683ea009e5.png)

**Reply by Satnet** [reply]
It does run, it wont rebuild tho because TempMon_40/41.hex is missing.
But yes, I can confirm that the console window opens and it reads the Micromod!

I did try to implement this into my own project, problem is that the UsbLib and Teensysharp Dev branch doesn't exist in Nuget.

What is your plans ahead for this project? :)

**Reply by luni64** [reply]
> What is your plans ahead for this project? :)

This is a seldom used library... Anyway, I'd like to get the new version going asap to get it off my desk :-) 

> I did try to implement this into my own project, problem is that the UsbLib and Teensysharp Dev branch doesn't exist in Nuget.

libUsbTee is on nuget, you need to tick  "prerelease versions" to find it. **But, usually you don't need to reference it since it is a dependency of TeensySharp.**

**Regarding the dev branch:**  Just git-clone TeensySharp to some folder (doesn't need to be in your solution folder). Then checkout the devlopment branch. If you are not familiar with git I recommend to install [Github Desktop](https://desktop.github.com/) which is really easy to use. 

From within your own solution right click on solution in the solution explorer, select 'add existing project', browse for the TeensySharp project and select it.  This should add  TeensySharp to your solution explorer. To use it from your project you now need to add a project reference (right click on your project, select add | project reference and tick TeensySharp)

Sounds more complicated than it really is.

**Reply by Satnet** [reply]
I managed to get the library into my own project, referencing libUsb as a DLL (uggly hack). Alltho console wont open in my project but that's nothing to care about. I do get a list of connected devices (with your test file) so that's all good. It wont auto update but that's once again due to my project, it did auto update in your example file.

Now it's just about the actual flashing procedure. :)

**Reply by Satnet** [reply]
Luni, May I ask how the progress is going for the download part of the script? As you know, Teensyduino 1.54 needs something else to work, it will not go into program mode automatically.

**Comment by Satnet** [comment]
Wonderful! I will try this tomorrow and let you know!



sön 11 juli 2021 kl. 20:53 skrev luni64 ***@***.***>:

**Comment by Satnet** [comment]
Hello Luni,

I hope you are well?
TeensySharp for the MM, I understand that the base of people using it is
probably small. But I am prepared to pay you to finish the library so that
I can use it to flash my MM.

Can we work something out?






mån 12 juli 2021 kl. 13:42 skrev luni64 ***@***.***>:

**Comment by luni64** [comment]
No need for payment.  

A couple of users are using it now and I didn't hear of any issues.  Also, the last thing I heard from you is that you wanted to try it and report errors? What was the result?  What exactly doesn't work and need to be fixed?

**Comment by Satnet** [comment]
Hello Luni,

I’m sorry for the confusion. I don’t really remember what the last thing I
said was.

I will download it and try, thanks!!

mån 11 okt. 2021 kl. 12:53 skrev luni64 ***@***.***>:

**Comment by luni64** [comment]
Great, let me know if it works. I'll push it to nuget then.

**Comment by Satnet** [comment]
Wonderful, I will let you know how it goes. Thanks man!

Den mån 11 okt. 2021 kl 12:56 skrev luni64 ***@***.***>:

**Comment by Satnet** [comment]
Luni,

When trying to download the dev version in Nuget, it doesn't exist.
This version was uploaded 2019 so it can't be that one?

See attached pic :)

Den mån 11 okt. 2021 kl 13:00 skrev Patrik ***@***.***>:

**Comment by luni64** [comment]
I meanwhile uploaded a beta version to nuget. It currently is called **libTeensySharp**  version: 0.9.1-beta https://www.nuget.org/packages/libTeensySharp/0.9.1-beta.  For now, I changed the references in the GitHub examples to nuget. -> The examples should run from the nuget version, not the one in the repo.   

I recently added the example "WinForms Test" which is a simple firmware uploader. I tested it with MM and it works OK. Let me know if it works for you.

**Comment by Satnet** [comment]
Luni,

I have downloaded the Dev lib from Nuget, and I have a list of USB devices,
so far so good. I'm not using the console tho, I don't really know how to
open it. I think it should open automatically. However, I have a rich text
box which I append text to.
In my .cs page in the public function on top, I have your print&watch code.
Alltho it doesn't update when I plug/unplug devices. I have tried with both
thread sleep and without.

Also, to upload, should I use this example:
https://github.com/luni64/TeensySharp/blob/develop/src/Examples/Firmware%20Upload%20Console/Program.cs
?
It seems wrong as it uses the old TeensyWatcher() which is not recognized
by Visual Studio.

Thanks!


Den mån 11 okt. 2021 kl 18:26 skrev luni64 ***@***.***>:

**Comment by luni64** [comment]
> I'm not using the console tho, I don't really know how to
open it. I think it should open automatically. However, I have a rich text
box which I append text to.
In my .cs page in the public function on top, I have your print&watch code.
Alltho it doesn't update when I plug/unplug devices. I have tried with both
thread sleep and without.

Don't really understand what you are saying here. 

**Instruction:** 
Please download the repository from the development branch. The /src folder contains a solution which you need to open with VS. The solution contains various examples which should run out of the box (in the VS solution explorer right click on an example project and set it as default startup project, then F5 should start the project), The  example "WinForms Test" is a fully functional downloader which shows how to download with the new lib. If this is running and you understand how  it works, switch to your own solution and implement it there. 

> It seems wrong as it uses the old TeensyWatcher() which is not recognized
by Visual Studio.

Yes, I didn't adopt the readme for the new library. For the time being, just use the examples in the solution as documentation.

**Comment by Satnet** [comment]
Luni,

Progress is good. I'm using your WiunForms Test example which works
perfectly. I'm going to adopt that to my design. Thank you!

Den tis 12 okt. 2021 kl 12:24 skrev luni64 ***@***.***>:
